### PR TITLE
Only stack allocate when marshalling Utf8 arguments

### LINKED ISSUE #74394: UTF-8 marshalling of string fields is broken
Hit this while running the triangle sample at https://github.com/MichalStrehovsky/sokol-csharp/blob/master/src/examples/triangle.cs.

```
 	triangle.exe!S_P_CoreLib_Interop_Ole32__CoTaskMemFree()	Unknown
 	triangle.exe!S_P_CoreLib_System_Runtime_InteropServices_Marshal__FreeCoTaskMem() Line 181	Unknown
 	triangle.exe!S_P_CoreLib_System_Runtime_InteropServices_Marshalling_Utf8StringMarshaller__Free() Line 49	Unknown
 	triangle.exe!Internal_CompilerGenerated__Module___<Cleanup>sokol_csharp_unofficial_Sokol_Gfx_BufferDesc()	Unknown
 	triangle.exe!sokol_csharp_unofficial_Sokol_Gfx__MakeBuffer_1()	Unknown
 	triangle.exe!sokol_csharp_unofficial_Sokol_Gfx__MakeBuffer_1()	Unknown
 	triangle.exe!sokol_csharp_unofficial_Sokol_Gfx__MakeBuffer_0<Single>()	Unknown
>	triangle.exe!triangle_Program____Main___g__Init_0_0() Line 33	Unknown
 	[Inline Frame] triangle.exe!_sapp_call_init() Line 2322	C
 	triangle.exe!_sapp_frame() Line 2510	C
 	triangle.exe!_sapp_win32_run(const sapp_desc * desc) Line 6594	C
 	triangle.exe!sokol_csharp_unofficial_Sokol_App__Run()	Unknown
 	triangle.exe!triangle_Program___Main__() Line 6	Unknown
```

CoTaskMemFree is detecting corrupted memory. Stepping through the marshalling code, we never allocated any because it was stack allocated. To add insult to injury, it was also stack allocated in a frame that we left and overwrote.

I think this regressed in #69043 that added the optimized path without allocation. We need to not do the optimized path when marshalling an element of a struct.

We need to start running the Pri-0 interop tests ASAP. Now that the runtime repo completely migrated off DllImport, there's nothing in the libraries partition that would be testing built in marshalling and our smoke tests are lacking coverage.

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
Fixes #74394.

Cc @dotnet/ilc-contrib

### CODE REVIEW

File: src/tests/nativeaot/SmokeTests/PInvoke/PInvokeNative.cpp [hunk 1]
```diff
@@ -529,14 +540,18 @@ DLL_EXPORT void __stdcall StructTest_ByOut(NativeSequentialStruct *nss)
 
     int arrSize = 7;
     char *p;
-    p = (char *)MemAlloc(sizeof(char) * arrSize);
+    p = (char *)MemAlloc(sizeof(char) * arrSize + 1);
```
Comment (MichalStrehovsky) [inline src/tests/nativeaot/SmokeTests/PInvoke/PInvokeNative.cpp:543]: This is unrelated. I almost copy pasted this so I looked at it. We were writing the null terminator past the end of the buffer...

File: src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.cs [hunk 2]
```diff
@@ -1672,7 +1672,7 @@ protected override void TransformManagedToNative(ILCodeStream codeStream)
         {
             ILEmitter emitter = _ilCodeStreams.Emitter;
 
-            if (In && !Out && !IsManagedByRef)
+            if (In && !Out && !IsManagedByRef && MarshallerType == MarshallerType.Argument)
```
Comment (jkotas) [inline src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.cs:1675]: I would expect this condition to be identical to:

https://github.com/dotnet/runtime/blob/170635417c3a87afd21c6138bdbdf2fd2874148b/src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.cs#L1476-L1480

Comment (jkotas) [reply src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.cs:1675]: Maybe both of these places should be changed to:
```
MarshalDirection == MarshalDirection.Forward 
     && MarshallerType == MarshallerType.Argument 
     && !IsManagedByRef 
     && In 
     && !Out
```
?

**Review by jkotas — APPROVED** [review]
Thank you for fixing this!

**Pull request merged**